Mathlib/Algebra/Module/LinearMap/Polynomial.lean

```diff
@@ -0,0 +1,371 @@
+    Finsupp.equivFunOnFinite, Equiv.coe_fn_symm_mk, Finsupp.coe_mk]
+
+open Algebra.TensorProduct TensorProduct in
+lemma polyCharpolyAux_map_aeval
+    (A : Type*) [CommRing A] [Algebra R A] [Module.Finite A (A ⊗[R] M)] [Module.Free A (A ⊗[R] M)]
+    (x : ι → A) :
+    (polyCharpolyAux φ b bₘ).map (MvPolynomial.aeval x).toRingHom =
+      LinearMap.charpoly ((tensorProduct R A M M).comp (baseChange A φ)
+        ((basis A b).repr.symm (Finsupp.equivFunOnFinite.symm x))) := by
+  rw [← polyCharpolyAux_map_eval (tensorProduct R A M M ∘ₗ baseChange A φ) _ (basis A bₘ),
+    polyCharpolyAux_baseChange, Polynomial.map_map]
+  congr
+  exact DFunLike.ext _ _ fun f ↦ (MvPolynomial.eval_map (algebraMap R A) x f).symm
+
+open Algebra.TensorProduct MvPolynomial in
+/-- `LinearMap.polyCharpolyAux` is independent of the choice of basis of the target module.
+
+Proof strategy:
+1. Rewrite `polyCharpolyAux` as the (honest, ordinary) characteristic polynomial
+   of the basechange of `φ` to the multivariate polynomial ring `MvPolynomial ι R`.
+2. Use that the characteristic polynomial of a linear map is independent of the choice of basis.
+   This independence result is used transitively via
+   `LinearMap.polyCharpolyAux_map_aeval` and `LinearMap.polyCharpolyAux_map_eq_charpoly`. -/
+lemma polyCharpolyAux_basisIndep {ιM' : Type*} [Fintype ιM'] [DecidableEq ιM']
+    (bₘ' : Basis ιM' R M) :
+    polyCharpolyAux φ b bₘ = polyCharpolyAux φ b bₘ' := by
+  let f : Polynomial (MvPolynomial ι R) → Polynomial (MvPolynomial ι R) :=
+    Polynomial.map (MvPolynomial.aeval X).toRingHom
+  have hf : Function.Injective f := by
+    simp only [f, aeval_X_left, AlgHom.toRingHom_eq_coe, AlgHom.id_toRingHom, Polynomial.map_id]
+    exact Polynomial.map_injective (RingHom.id _) Function.injective_id
+  apply hf
+  let _h1 : Module.Finite (MvPolynomial ι R) (TensorProduct R (MvPolynomial ι R) M) :=
+    Module.Finite.of_basis (basis (MvPolynomial ι R) bₘ)
+  let _h2 : Module.Free (MvPolynomial ι R) (TensorProduct R (MvPolynomial ι R) M) :=
+    Module.Free.of_basis (basis (MvPolynomial ι R) bₘ)
+  simp only [f, polyCharpolyAux_map_aeval, polyCharpolyAux_map_aeval]
+
+end aux
+
+open FiniteDimensional Matrix
+
+variable [Module.Free R M] [Module.Finite R M] (b : Basis ι R L)
+
+/-- Let `L` and `M` be finite free modules over `R`,
+and let `φ : L →ₗ[R] Module.End R M` be a linear map.
+Let `b` be a basis of `L` and `bₘ` a basis of `M`.
+Then `LinearMap.polyCharpoly φ b` is the polynomial that evaluates on elements `x` of `L`
+to the characteristic polynomial of `φ x` acting on `M`. -/
+noncomputable
+def polyCharpoly : Polynomial (MvPolynomial ι R) :=
+  φ.polyCharpolyAux b (Module.Free.chooseBasis R M)
+
+lemma polyCharpoly_eq_polyCharpolyAux [DecidableEq ιM] (bₘ : Basis ιM R M) :
+    polyCharpoly φ b = φ.polyCharpolyAux b bₘ := by
+  rw [polyCharpoly, φ.polyCharpolyAux_basisIndep b (Module.Free.chooseBasis R M) bₘ]
```

Comment:
I think we should regard `polyCharpolyAux` as a pure implementation detail and never expose it to the user so this lemma should probably be:
```suggestion
lemma polyCharpoly_eq_of_basis [DecidableEq ιM] (bₘ : Basis ιM R M) :
    polyCharpoly φ b =
    (charpoly.univ R ιM).map (MvPolynomial.bind₁ (φ.toMvPolynomial b bₘ.end)) := by
  rw [polyCharpoly, φ.polyCharpolyAux_basisIndep b (Module.Free.chooseBasis R M) bₘ,
    polyCharpolyAux]
```